### PR TITLE
Implement DiagnosticDataProvider interface

### LINKED ISSUE #11922: Introduce DiagnosticDataProvider + DiagnosticsDelegate for all diagnostics clusters.
#### Problem
Per sync offline, we will introduce DiagnosticDataProvider + DiagnosticsDelegate for all diagnostics clusters. Only handle these three attributes within this PR with new interface and will refactor all other diagnostics attributes to use this new interface in the followup PRs

 #### Proposed Solution
<suggested fix, suggested enhancement>

### PR DESCRIPTION
**Pull request by yufengwangca**
#### Problem
What is being fixed?  Examples:
* Current exiting platform interfaces based on CRTP are complicated and over-bloated. Instead of plugging all operations to existing APIs, we will introduce dedicated DiagnosticDataProvider + DiagnosticsDelegate platform interface for all diagnostics clusters, and migrate all diagnostics attributes to use this new interface.

* Fixes #11922

#### Change overview
Implement DiagnosticDataProvider interface

Diagnostics attributes in ConnectivityMgr will be migrated in another PR.

#### Testing
How was this tested? (at least one bullet point required)
* Confirm the all existing diagnostic attributes read still functional after the refactor.
```
yufengw@yufengw-SEi:~/connectedhomeip/out/debug/standalone$ ./chip-tool generaldiagnostics read reboot-count 12344321 0
.....
[1637218104.215705][158246:158251] CHIP:EM: Removed CHIP MessageCounter:1 from RetransTable on exchange 65295i
[1637218104.215719][158246:158251] CHIP:DMG: ReportDataMessage =
[1637218104.215727][158246:158251] CHIP:DMG: {
[1637218104.215732][158246:158251] CHIP:DMG: 	AttributeReportIBs =
[1637218104.215739][158246:158251] CHIP:DMG: 	[
[1637218104.215746][158246:158251] CHIP:DMG: 		AttributeReportIB =
[1637218104.215757][158246:158251] CHIP:DMG: 		{
[1637218104.215764][158246:158251] CHIP:DMG: 			AttributeDataIB =
[1637218104.215780][158246:158251] CHIP:DMG: 			{
[1637218104.215790][158246:158251] CHIP:DMG: 				AttributePathIB =
[1637218104.215795][158246:158251] CHIP:DMG: 				{
[1637218104.215800][158246:158251] CHIP:DMG: 					Endpoint = 0x0,
[1637218104.215804][158246:158251] CHIP:DMG: 					Cluster = 0x33,
[1637218104.215809][158246:158251] CHIP:DMG: 					Attribute = 0x0000_0001,
[1637218104.215813][158246:158251] CHIP:DMG: 				}
[1637218104.215818][158246:158251] CHIP:DMG: 					
[1637218104.215822][158246:158251] CHIP:DMG: 					Data = 1, 
[1637218104.215826][158246:158251] CHIP:DMG: 				DataVersion = 0x0,
[1637218104.215830][158246:158251] CHIP:DMG: 			},
[1637218104.215835][158246:158251] CHIP:DMG: 			
[1637218104.215840][158246:158251] CHIP:DMG: 		},
[1637218104.215847][158246:158251] CHIP:DMG: 		
[1637218104.215853][158246:158251] CHIP:DMG: 	],
[1637218104.215860][158246:158251] CHIP:DMG: 	
[1637218104.215863][158246:158251] CHIP:DMG: }
[1637218104.215883][158246:158251] CHIP:ZCL: ReadAttributesResponse:
[1637218104.215886][158246:158251] CHIP:ZCL:   ClusterId: 0x0000_0033
[1637218104.215890][158246:158251] CHIP:ZCL:   attributeId: 0x0000_0001
[1637218104.215893][158246:158251] CHIP:ZCL:   status: Success                (0x0000)
[1637218104.215896][158246:158251] CHIP:ZCL:   attribute TLV Type: 0x04
[1637218104.215899][158246:158251] CHIP:TOO: Int16u attribute Response: 1
```

### CODE REVIEW

File: src/platform/Ameba/DiagnosticDataProviderImpl.h [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+#include <platform/DiagnosticDataProvider.h>
+
+namespace chip {
+namespace DeviceLayer {
+
+/**
+ * Concrete implementation of the PlatformManager singleton object for Linux platforms.
+ */
+class DiagnosticDataProviderImpl final : public DiagnosticDataProvider
```
Comment (tcarmelveilleux) [inline src/platform/Ameba/DiagnosticDataProviderImpl.h:33]: Not sure these need to be final

File: src/platform/Ameba/DiagnosticDataProviderImpl.h [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+{
+public:
+    static DiagnosticDataProviderImpl & GetDefaultInstance();
+
+private:
+    // ===== Methods that implement the PlatformManager abstract interface.
```
Comment (tcarmelveilleux) [inline src/platform/Ameba/DiagnosticDataProviderImpl.h:39]: I am confused why we change these to private

File: src/platform/Ameba/DiagnosticDataProviderImpl.h (hunk 1 above)
Comment (yufengwangca) [reply src/platform/Ameba/DiagnosticDataProviderImpl.h:33]: We already reach to the end of platform implementation, it should be final at this layer just like other classes in this folder

File: src/platform/Ameba/DiagnosticDataProviderImpl.h (hunk 2 above)
Comment (yufengwangca) [reply src/platform/Ameba/DiagnosticDataProviderImpl.h:39]: Other classes put implementation parts as private except new added ThreadStackManagerImpl, since those implementation functions are not be used outside directly, is there any reasons to make it pubic?

File: src/platform/EFR32/DiagnosticDataProviderImpl.h [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+/**
+ * Concrete implementation of the PlatformManager singleton object for Linux platforms.
+ */
+class DiagnosticDataProviderImpl final : public DiagnosticDataProvider
+{
+public:
+    static DiagnosticDataProviderImpl & GetDefaultInstance();
+
+private:
+    // ===== Methods that implement the PlatformManager abstract interface.
```
Comment (tcarmelveilleux) [inline src/platform/EFR32/DiagnosticDataProviderImpl.h:41]: Please leave all methods public so that a product can override 1-2 methods  from the platform default easily

File: src/platform/Ameba/DiagnosticDataProviderImpl.h (hunk 1 above)
Comment (yufengwangca) [reply src/platform/Ameba/DiagnosticDataProviderImpl.h]: Removed final to allow product overwrite

File: src/platform/Ameba/DiagnosticDataProviderImpl.h (hunk 2 above)
Comment (yufengwangca) [reply src/platform/Ameba/DiagnosticDataProviderImpl.h]: Change to public to allow product overwrite

File: src/include/platform/DiagnosticDataProvider.h [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+    DiagnosticDataProvider(const DiagnosticDataProvider &&) = delete;
+    DiagnosticDataProvider & operator=(const DiagnosticDataProvider &) = delete;
+};
+
+/**
+ * Returns a reference to a DiagnosticDataProvider object.
+ *
+ * Applications should use this to access the features of the DiagnosticDataProvider.
+ */
+extern DiagnosticDataProvider & DiagnosticDataMgr();
```
Comment (tcarmelveilleux) [inline src/include/platform/DiagnosticDataProvider.h:111]: This does not need to say `extern`.

Also, it would be good for symettry to call this `SetDiagnosticDataProvider()`

Comment (yufengwangca) [reply src/include/platform/DiagnosticDataProvider.h]: 'extern' is optional, we can remove, just align the style with other platform interface classes.

We need to use different name than constructor name, just as use PlatformMgr for PlatformManager class.

File: src/include/platform/ConfigurationManager.h [hunk 5]
```diff
@@ -102,8 +102,8 @@ class ConfigurationManager
     virtual CHIP_ERROR StoreRebootCount(uint32_t rebootCount)                          = 0;
     virtual CHIP_ERROR GetTotalOperationalHours(uint32_t & totalOperationalHours)      = 0;
     virtual CHIP_ERROR StoreTotalOperationalHours(uint32_t totalOperationalHours)      = 0;
-    virtual CHIP_ERROR GetBootReasons(uint32_t & bootReasons)                          = 0;
-    virtual CHIP_ERROR StoreBootReasons(uint32_t bootReasons)                          = 0;
+    virtual CHIP_ERROR GetBootReason(uint32_t & bootReason)                            = 0;
+    virtual CHIP_ERROR StoreBootReason(uint32_t bootReason)                            = 0;
```
Comment (tcarmelveilleux) [inline src/include/platform/ConfigurationManager.h:106]: These setters should not be COnfigurationManager anymore, only on DiagnosticsDataProvider. Otherwise we have an extra dependency.

This can be refactored later.

File: src/include/platform/DiagnosticDataProvider.h (hunk 4 above)
Comment (tcarmelveilleux) [reply src/include/platform/DiagnosticDataProvider.h]: > We need to use different name than constructor name, just as use PlatformMgr for PlatformManager class.

I do not understand this comment.

The naming here is confusing. The class is called `DiagnosticsDataProvider`. Why is the getter for a singleton isntance called `DiagnosticsDataMgr` ?

File: src/include/platform/ConfigurationManager.h (hunk 5 above)
Comment (yufengwangca) [reply src/include/platform/ConfigurationManager.h:106]: Those values, such as BootReason could be set before the device get rebooted, and this value need to survive through the power cycle in this case. Any values need to survive through the power cycle are currently stored in ConfigurationMgr

**Review by tcarmelveilleux — APPROVED** [review]

File: src/include/platform/DiagnosticDataProvider.h (hunk 4 above)
Comment (yufengwangca) [reply src/include/platform/DiagnosticDataProvider.h]: Suppose we have following singleton class which use same name for constructor and getter.

class Singleton {
public:
    static Singleton * Singleton(); // returns the unique instance
    static void destroyInstance(); // returns the unique instance
    void print();
    
private:
    static Singleton *instance;
    Singleton(); // Private constructor: a singleton instance is accessible only through instance
    ~Singleton();
    Singleton(const Singleton&); // prevent compiler-generated copy constructor
    Singleton& operator=(const Singleton&); // prevent compiler-generated operator =
};

I will get the following compile error:

singleton.cpp:11:34: error: return type specification for constructor invalid
   11 |     static Singleton * Singleton(); // returns the unique instance

**Review by jmartinez-silabs — APPROVED** [review]

File: src/include/platform/DiagnosticDataProvider.h [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+    virtual CHIP_ERROR GetCurrentHeapFree(uint64_t & currentHeapFree);
+    virtual CHIP_ERROR GetCurrentHeapUsed(uint64_t & currentHeapUsed);
+    virtual CHIP_ERROR GetCurrentHeapHighWatermark(uint64_t & currentHeapHighWatermark);
+
+    /*
+     * Get the linked list of thread metrics of the current plaform. After usage, each caller of GetThreadMetrics
+     * needs to release the thread metrics list it gets via ReleaseThreadMetrics.
+     *
+     */
+    virtual CHIP_ERROR GetThreadMetrics(ThreadMetrics ** threadMetricsOut);
```
Comment (Damian-Nordic) [inline src/include/platform/DiagnosticDataProvider.h:89]: `ForEachThread(void (*visitor)(const ThreadInfo&, uintptr_t), uintptr_t context)` might be more scalable interface as some applications may use quite a lot of threads.

File: src/platform/Zephyr/DiagnosticDataProviderImpl.cpp [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+ */
+
+/**
+ *    @file
+ *          Provides an implementation of the DiagnosticDataProvider object
+ *          for Zephy platform.
+ */
+
+#if !CONFIG_NORDIC_SECURITY_BACKEND
+#include <crypto/CHIPCryptoPAL.h> // nogncheck
```
Comment (Damian-Nordic) [inline src/platform/Zephyr/DiagnosticDataProviderImpl.cpp:25]: Is it needed?

File: src/platform/Zephyr/DiagnosticDataProviderImpl.cpp [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+#endif                            // !CONFIG_NORDIC_SECURITY_BACKEND
+
+#include <platform/internal/CHIPDeviceLayerInternal.h>
+
+#include <lib/support/logging/CHIPLogging.h>
+#include <platform/DiagnosticDataProvider.h>
+#include <platform/Zephy/DiagnosticDataProviderImpl.h>
+
+#include <drivers/entropy.h>
```
Comment (Damian-Nordic) [inline src/platform/Zephyr/DiagnosticDataProviderImpl.cpp:34]: Seems unrelated.

File: src/platform/Zephyr/PlatformManagerImpl.cpp [hunk 9]
```diff
@@ -88,41 +88,5 @@ CHIP_ERROR PlatformManagerImpl::_InitChipStack(void)
     return err;
 }
 
-CHIP_ERROR PlatformManagerImpl::_GetCurrentHeapFree(uint64_t & currentHeapFree)
```
Comment (Damian-Nordic) [inline src/platform/Zephyr/PlatformManagerImpl.cpp:91]: Please add `SetDiagnosticDataProvider(&DiagnosticDataProviderImpl::GetDefaultInstance());` in `InitChipStack` method above.

File: src/platform/nrfconnect/BUILD.gn [hunk 10]
```diff
@@ -39,6 +39,8 @@ static_library("nrfconnect") {
     "ConnectivityManagerImpl.h",
     "DeviceNetworkProvisioningDelegateImpl.cpp",
     "DeviceNetworkProvisioningDelegateImpl.h",
+    "DiagnosticDataProviderImpl.cpp",
```
Comment (Damian-Nordic) [inline src/platform/nrfconnect/BUILD.gn:42]: nRF Connect SDK is based on Zephyr so it can by default use `../Zephyr/DiagnosticDataProviderImpl.cpp` and doesn't require its own implementation.

File: src/platform/telink/BUILD.gn [hunk 11]
```diff
@@ -39,6 +39,8 @@ static_library("telink") {
     "ConnectivityManagerImpl.h",
     "DeviceNetworkProvisioningDelegateImpl.cpp",
     "DeviceNetworkProvisioningDelegateImpl.h",
+    "DiagnosticDataProviderImpl.cpp",
```
Comment (Damian-Nordic) [inline src/platform/telink/BUILD.gn:42]: Telink is also a Zephyr-based platform, so it can use Zephyr's implementation.

**Review by Damian-Nordic — CHANGES_REQUESTED** [review]
Please make sure Zephyr's implementation is used in nRF connect and Telink platforms, and is initialized properly.

**Review by Damian-Nordic — APPROVED** [review]

File: src/include/platform/DiagnosticDataProvider.h (hunk 6 above)
Comment (yufengwangca) [reply src/include/platform/DiagnosticDataProvider.h:89]: I think this also apply to interface    
```
CHIP_ERROR GetNetworkInterfaces(NetworkInterface ** netifpp);
void ReleaseNetworkInterfaces(NetworkInterface * netifp);
```

What I would like to do is to add iterator to netifpp struct, so we can just use following syntax to iterate all available active interface.  

```
        for (auto & ifp : netifs)
        {

        }

```
I will do it in follow-up https://github.com/project-chip/connectedhomeip/issues/12106 and leave this PR only to refactor the current implemenation from PlatformMgr to DiagnosticDataProvider

Comment (Damian-Nordic) [reply src/include/platform/DiagnosticDataProvider.h:89]: Yeah, a custom iterator would also be nice. In case of network interfaces we already have `InterfaceIterator` in `InetInterface.h` which is implemented for most platforms, so it's worth considering using it instead of creating another mechanism.

Comment (Damian-Nordic) [reply src/include/platform/DiagnosticDataProvider.h:89]: But that's just an idea. I'm not requesting these changes as part of this PR.

**Review by saurabhst — APPROVED** [review]
